### ISSUE TITLE
Last result cache

### ISSUE THREAD

**Issue by Odrakir** [ReactiveX/RxSwift#324]
Hi, 

I have a time consuming operation that needs to be repeated every time the user makes some action. I want to be able to store the operation result so I can return the stored result when requested and start calculating the next result immediately.

This is what I have, it works, but it doesn't look like the best way to do it.

```swift
    var cached:MyResult?
    
    func getMyResult() -> Observable<MyResult>
    {
        
        if let cached = cached {
            self.cached = nil
            timeConsumingOperation()
                .retry(3)
                .subscribeNext {
                    response in
                    self.cached = response
                }
                .addDisposableTo(disposeBag)
            
            return just(cached)
        } else {
            return timeConsumingOperation()
                .retry(3)
                .doOn() {
                    response in
                    self.timeConsumingOperation()
                        .retry(3)
                        .subscribeNext {
                            self.cached = $0
                        }
                        .addDisposableTo(self.disposeBag)
            }
        }
    }
    
    func timeConsumingOperation() -> Observable<MyResult>
    {
        ...
    }
```

**Comment by yury** [comment]
```swift
var _cached:MyResult?

func getMyResult() -> Observable<MyResult> {
   let o = timeConsumingOperation()
   .retry(3)
   .doOn(onNext: { r in self._cached = r } )

   if let cached = _cached {
      o = o.startWith(cached)
   }

   return o
}
```

This is a little bit different... You will get two `.Next(result)` events. But first with cached if available and second with actual data.

**Comment by Odrakir** [comment]
That doesn't seem to do what I want:
- The first time I call the method, as there's no cached result it has to call the timeConsumingOperation twice, one for the current result, and another one for the cache.
- The successive times it has to return the cache result and call the operation again to generate the next cached result.

Your code just does one calculation the first time and in successive calls it returns the last value again and then returns and stores a new value.

**Comment by fpillet** [comment]
This will do what you want:

```
	var cache : MyResult?
	
	func timeConsumingOperation() -> Observable<MyResult> {
		// your long stuff here
	}
	
	func getMyResult() -> Observable<MyResult> {
		return create {
			observer in
			return (self.cache != nil ? just(self.cache!) : empty())
				.concat(self.timeConsumingOperation().doOn(onNext: { self.cache = $0 }))
				.elementAt(0)
				.subscribe(observer)
		}
	}
```

I left additional `[weak self]` shielding as an exercise to the reader :)

**Comment by kzaher** [comment]
I think this is a perfect time and place to advertise our `deferred` operator :)

```
var cache : MyResult?

func timeConsumingOperation() -> Observable<MyResult> {
    return just(MyResult()) // we can pretend this is time consuming
}

func getMyResult() -> Observable<MyResult> {
    return deferred {
        let start = cache.map(just) ?? empty()
        let suffix = timeConsumingOperation()
            .doOn(onNext: { cache = $0 })

        return start.concat(suffix)
    }
}
```

One additional hint, this is conceptual solution. Please synchronize this assignment `cache = $0` and reading of `cache.map(just)` because rx is free threaded. 

You can ofc get away without additional synchronization in case you know on which scheduler is subscription being invoked and operation result delivered.

**Comment by fpillet** [comment]
@kzaher nice trick with `cache.map(just)` I had already seen it but forgot about it.

This said, your solution still misses the `elementAt(0)`. The exact requirement expressed above was to get the last cached value, and immediately start a task to compute the next one (but apparently the idea it to only return one result and stop there, hence the `elementAt(0)`)

**Comment by fpillet** [comment]
Occurs to me that using a `Variable` would fit the locking requirement. Also note that additional work needs to be done if you want to guarantee that there is only one `timeConsumingOperation()` running at a time. This all depends on the kind of time-consuming processing that you're doing, and the expected behavior;

Hence:

```
var cache = Variable<MyResult?>(nil)

func timeConsumingOperation() -> Observable<MyResult> {
    return just(MyResult()) // we can pretend this is time consuming
}

func getMyResult() -> Observable<MyResult> {
    return deferred {
        let start = cache.value.map(just) ?? empty()
        let suffix = timeConsumingOperation()
            .doOn(onNext: { cache.value = $0 })

        return start.concat(suffix).elementAt(0)
    }
}
```

**Comment by Odrakir** [comment]
Can't get any of the solutions to work. Both @fpillet solutions with `elementAt(0)`always return the first result, over and over again. @kzaher solution doesn't return anything until the operation finishes.

**Comment by Odrakir** [comment]
Actually, what happens with @kzaher solution is, its sending the result when it finishes calculating it, even if it already sent the cached result. It shouldn't do that, it should store it as cached and send it the next time the method is called.

**Comment by fpillet** [comment]
I think you did something wrong. Last code I posted works just fine with the requirements you expressed, which are:
- Observable will return one value and complete (that's what you wanted, right?)
- A subscription to Observable will:
  * start a timeConsumingOperation() to compute the next result,
  * immediately return the last cached result if it exists, wait for timeConsumingOperation() to be done then complete, or
  * wait for timeConsumingOperation() to produce a result and send it then complete

I just tried it in the Rx playground and it works as expected:

```
var cache = Variable<Int?>(nil)
var ncache = 1
func timeConsumingOperation() -> Observable<Int> {
	return just(ncache++) // we can pretend this is time consuming
}

func getMyResult() -> Observable<Int> {
	return deferred {
		let start = cache.value.map(just) ?? empty()
		let suffix = timeConsumingOperation()
			.doOn(onNext: { cache.value = $0 })
		
		return start.concat(suffix).elementAt(0)
	}
}

_ = getMyResult().subscribeNext { print ("First subscription got \($0)") }
_ = getMyResult().subscribeNext { print ("Second subscription got \($0)") }
_ = getMyResult().subscribeNext { print ("Third subscription got \($0)") }
```

This produces:

```
First subscription got 1
Second subscription got 1
Third subscription got 2
```

First subscription has no cache, computes then return 1
Second subscription has cache, returns it (1) then computes
Third subscription has cache, returns it (2) then computes.

**Comment by Odrakir** [comment]
Actually, you are right, I always get the same result because I'm trying to execute the time consuming operation on a background thread.


``` swift
func getMyResult() -> Observable<Int> {
    return deferred {
        let start = cache.value.map(just) ?? empty()
        let suffix = timeConsumingOperation()
            .subscribeOn(OperationQueueScheduler(operationQueue: NSOperationQueue()))
            .doOn(onNext: { cache.value = $0 })

        return start.concat(suffix).elementAt(0)
    }
}
```

**Comment by Odrakir** [comment]
Don't really know what I'm doing, but this seems to fix it:
```swift
func getMyResult() -> Observable<Int> {
    return deferred {
        let start = cache.value.map(just) ?? empty()
        let suffix = timeConsumingOperation()
            .doOn(onNext: { cache.value = $0 })

        return start.concat(suffix).elementAt(0)
            .subscribeOn(OperationQueueScheduler(operationQueue: NSOperationQueue()))
    }
}
```

**Comment by fpillet** [comment]
Yes that's what I mentioned earlier:

> Also note that additional work needs to be done if you want to guarantee that there is only one timeConsumingOperation() running at a time. This all depends on the kind of time-consuming processing that you're doing, and the expected behavior;

**Comment by Odrakir** [comment]
Well, I'm really not sure about the Scheduler part, and it would be great if someone could shed some light about what's going on, but this seems to do exactly what I want:

```swift
    func getMyResult() -> Observable<MyResult>
    {
     
        return deferred {[unowned self] in
            let start = self.cached.map(just) ?? empty()
            let operation = self. timeConsumingOperation()
                .retry(3)
                .subscribeOn(OperationQueueScheduler(operationQueue: NSOperationQueue()))

            operation
                .subscribeNext {[unowned self] v in self.cached = v }
                .addDisposableTo(self.disposeBag)
            
            return start
                .concat(operation).elementAt(0)

        }
```

**Comment by Odrakir** [comment]
By the way, thanks @fpillet, I think the `concat().elementAt(0)` was the part that I was missing.

I've replaces `doOn` with `subscribeNext`because this way the cache always gets replaced. If there's no cache the operation will be done twice, if there's cache, only once.

**Comment by fpillet** [comment]
There's something fishy with your `timeConsumingOperation()` call. My code above (try it in the playground) makes it clear that the operation (which returns an observable) executes only once.

I added some additional prints in the code at various points and I get correct results:

```
getMyResult() called
timeConsumingOperation() called and will produce value 1
First subscription got 1
getMyResult() called
timeConsumingOperation() called and will produce value 2
Second subscription got 1
getMyResult() called
timeConsumingOperation() called and will produce value 3
Third subscription got 2
```

**Comment by Odrakir** [comment]
Yes, your code executes it only once. But the first time you call it you don't fill the cache, so the second time you return the first result again, that's something I wanted to avoid.

So that's why in my list code I used `subscribeNext`. If there's cache, that's what the call will return, so there'll be only one subscription to the operation (the outside subscription). If there isn't cache, it'll return the operation, thus there'll be two subscriptions, one for the current result and one for the cache.

That works as expected.

**Comment by fpillet** [comment]
Hrm what you're saying now contradicts what you originally asked for, which is what I provided an implementation for:

>  I want to be able to store the operation result so I can return the stored result when requested and start calculating the next result immediately.

Which is exactly what my code does. Therefore what you need is a tweak where if the cache is originally empty, then you need to make two calls to `timeConsumingResult()`: a first one to return a result for the requested observable, a second one to fill cache the result of the next request.

Anyway, you seem to be on track with this now. I think we can close the issue.

**Comment by Odrakir** [comment]
Just one more question: I've removed the deferred part from my last code and it seems to be doing the exact same thing. What is that for? The docs on reactivex.io say it defers the creation of the observable until subscription, but as I'm not really creating an observable there, just composing them, I guess is not needed ??

**Comment by fpillet** [comment]
What `deferred` is simple: take a factory function which produces an `Observable`, and call this function every time a new observer subscribes, in order to potentially return a new `Observable` to each subscriber.

I'll refer you to ReactiveX's documentation: http://reactivex.io/documentation/operators/defer.html and read the link pointed to as well.

**Comment by fpillet** [comment]
Also, every time you are composing observables you are effectively creating a new observable. Whether you return the same one to every new subscriber, or return a new one to each depends on your needs.

**Comment by Odrakir** [comment]
Ok, I think I understand. That doesn't explain some of the things going on in my code, but I think I'll leave it for another issue. Thanks.

**Comment by whlsxl** [comment]
Hi @fpillet 
A question is so confuse me.
Why use `var cache = Variable<Int?>(nil)` not just store `var cache = Int?`
And when don't just 

`if let cache = cache {
    return .just(cache)
 }`
after that return the http request

**Comment by fpillet** [comment]
@whlsxl `Variable` uses atomic access to the value (check the code). It's less code to use a variable for the cache than to add the lock statements manually :)

**Comment by whlsxl** [comment]
@fpillet  Thanks a lot!!!
And I have a second question...
could I just `if cache.value != nil { return cache.asObservable() }` instead of `concat` two `Observable`.

**Comment by BrianSemiglia** [comment]
If there was a `flatScan` operator you might be able to do something like this (pseudo):

```
Observable
    .just(Array.init)
    .flatScan { new, sum in
         sum.contains ? .just(sum) : async(new).map { $0 + sum }
     }
     .first {
        $0 == predicate
     }
```
